### PR TITLE
Add public key and chaincode calculation utility

### PR DESCRIPTION
**Pull request by fbeutin-ledger**

### CODE REVIEW

**Review by xchapron-ledger — DISMISSED** [review]

**Comment by xchapron-ledger** [comment]
Maybe the CI fix due to boilerplate artifact change could be put in another commit or another PR?

File: src/ragger/utils/bip/seed.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    CurveChoice.Ed25519Kholaw: Bip32Ed25519Kholaw,
+    CurveChoice.Ed25519Blake2bSlip: Bip32Ed25519Blake2bSlip,
+}
+
+SPECULOS_MNEMONIC = (
+    "glory promote mansion idle axis finger extra february uncover one trip resource "
+    "lawn turtle enact monster seven myth punch hobby comfort wild raise skin")
+
+
+def calculate_public_key_and_chain_code(curve: CurveChoice,
```
Comment (xchapron-ledger) [inline src/ragger/utils/bip/seed.py:33]: Maybe you can add a test for this function?

Comment (fbeutin-ledger) [reply src/ragger/utils/bip/seed.py]: Done

File: src/ragger/bip/seed.py [hunk 2]
```diff
@@ -1,86 +1,45 @@
...
+SPECULOS_MNEMONIC = (
+    "glory promote mansion idle axis finger extra february uncover one trip resource "
+    "lawn turtle enact monster seven myth punch hobby comfort wild raise skin")
+
+
+def calculate_public_key_and_chaincode(curve: CurveChoice,
+                                       path: str,
+                                       mnemonic=SPECULOS_MNEMONIC) -> Tuple[str, str]:
+    if not isinstance(curve, CurveChoice):
+        raise ValueError(f'{"curve must be a CurveChoice enum"}')
```
Comment (lpascal-ledger) [inline src/ragger/bip/seed.py:37]: Shouldn't it be `raise ValueError(f'"{curve}" must be a CurveChoice enum')` ?

File: src/ragger/utils/__init__.py [hunk 3]
```diff
@@ -15,11 +15,8 @@
...
 from .structs import RAPDU, Crop
 from .packing import pack_APDU
-from .path import BtcDerivationPathFormat, pack_derivation_path
-from .path import bitcoin_pack_derivation_path
 from .misc import app_path_from_app_name, prefix_with_len
 from .misc import create_currency_config, split_message
 
-__all__ = ("RAPDU", "pack_APDU", "Crop", "BtcDerivationPathFormat", "pack_derivation_path",
-           "bitcoin_pack_derivation_path", "app_path_from_app_name", "prefix_with_len",
+__all__ = ("RAPDU", "pack_APDU", "Crop", "app_path_from_app_name", "prefix_with_len",
```
Comment (lpascal-ledger) [inline src/ragger/utils/__init__.py:21]: `sort`

File: tests/bip/test_path.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+
+class TestPath(TestCase):
+
+    def n_level_test(self, n: int):
+        for variant in [0, 255, 81845, 887587, 2147483647]:
+            for hardening in [False, True]:
+                hardening_char = '\'' if hardening else ''
+                previous_levels = "0/" * (n - 1)
+                path = "m/{}{}{}".format(previous_levels, variant, hardening_char)
```
Comment (lpascal-ledger) [inline tests/bip/test_path.py:18]: Why using a `str.format` here, but using `f-strings` everywhere else?

File: tests/bip/test_seed.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    def test_get_public_key_and_chaincode_error_unhardened_path(self):
+        with self.assertRaises(Bip32KeyError):
+            calculate_public_key_and_chaincode(curve=CurveChoice.Ed25519Slip,
+                                               path="m/44'/601/1430/20/0/0")
+        with self.assertRaises(Bip32KeyError):
+            calculate_public_key_and_chaincode(curve=CurveChoice.Ed25519Blake2bSlip,
+                                               path="m/44'/601/1430/20/0/0")
+
+    def test_get_public_key_and_chaincode_secp256k1(self):
+        secp256k1_test_values: List[Tuple[str, str, str]] = (
```
Comment (lpascal-ledger) [inline tests/bip/test_seed.py:25]: Technically, this is a `Tuple[Tuple[str, str, str]]` but we don't really care, we just need iterables.

File: tests/bip/test_seed.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+             "4103e5fb91a143a23db9a210e698c196d146f7e7cad678d598bacf80efe7e28d"),
+            ("m/44'/0'/0'/0'/0'",
+             "00c2ea5e68ab428901e21cb1370765b4cff3abea905d4b4b2efebf977c6202782a",
+             "b40ac81f3f3a3edc493edc95fcb2a7b1735aa2a5a706f0ee4511c3d0f5dfaf06"),
+            ("m/44'/601'/1430'/20'/0'/0'",
+             "0059e258339ae430ded97912ed077038edf189ceb8f7d022741a9000950b131f30",
+             "a6e90bed1450e77a6f3933b075fa23be315e676ccd5eddd778a62e3d669ab2d6"),
+        )
+        for path, ref_public_key, ref_chaincode in ed25519slip_test_values:
+            pk, cc = calculate_public_key_and_chaincode(curve=CurveChoice.Ed25519Slip,
```
Comment (lpascal-ledger) [inline tests/bip/test_seed.py:79]: I feel like these 5 tests are the same, except for the input curve and the function result. Couldn't they be factorized / parametrized?

**Review by lpascal-ledger — CHANGES_REQUESTED** [review]
Not much to say, appart from cosmetic details

**Review by xchapron-ledger — DISMISSED** [review]
You deleted some utils. Do you know if there was use by anyone? I personnaly din't use them and find the one you introduced/cleanup complete enough.

File: src/ragger/__init__.py [hunk 7]
```diff
@@ -29,4 +31,8 @@
 
 logger.addHandler(handler)
 
-__all__ = ["Crop", "ExceptionRAPDU", "Firmware", "logger", "RAPDU", "SDK_VERSIONS"]
+__all__ = [
+    "ExceptionRAPDU", "Firmware", "SDK_VERSIONS", "RAPDU", "Crop", "logger",
```
[the remark this answers is not in the document]
Comment (fbeutin-ledger) [reply src/ragger/__init__.py:35]: Ah, I was not a fan of adding the exports but I did not want to change the way of doing

File: src/ragger/bip/path.py [hunk 8]
```diff
@@ -13,51 +13,35 @@
...
 
-    def __init__(self, elems: Optional[Union[bytes, Bip32PathArgs]] = None):
-        if not elems:
-            raise ValueError("A Bip32 path can't be empty")
-        if isinstance(elems, bytes):
-            super().__init__((elems[i:i + 4] for i in range(0, len(elems), 4)), is_absolute=True)
+    path_bytes: bytes = (len(split) - 1).to_bytes(1, byteorder='big')
+    for i in split[1:]:
+        if i == "":
+            raise ValueError("Error no value")
```
[the remark this answers is not in the document]
Comment (fbeutin-ledger) [reply src/ragger/bip/path.py:37]: Changed, new output:
`ValueError: Error no value among split list ['m', '', '0'] at index 0`

File: src/ragger/bip/path.py [hunk 9]
```diff
@@ -13,51 +13,35 @@
...
-
-    @property
-    def address_index(self) -> int:
-        return self._to_int(self[4])
+            path_bytes += int(i).to_bytes(4, byteorder='big')
+    return path_bytes
+
+
+def bitcoin_pack_derivation_path(format: BtcDerivationPathFormat, derivation_path: str) -> bytes:
+    assert isinstance(format, BtcDerivationPathFormat)
```
[the remark this answers is not in the document]
Comment (fbeutin-ledger) [reply src/ragger/bip/path.py:46]: Using raise instead of assert

File: tests/bip/test_seed.py (hunk 6 above)
Comment (fbeutin-ledger) [reply tests/bip/test_seed.py:79]: Done, as much as I could (not that much)

File: tests/bip/test_path.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
+from unittest import TestCase
+
+from ragger import pack_derivation_path
+
+MAX_VALUE = 2**31 - 1
```
[the remark this answers is not in the document]
Comment (fbeutin-ledger) [reply tests/bip/test_path.py:5]: Now yes !

**Comment by fbeutin-ledger** [comment]
> You deleted some utils. Do you know if there was use by anyone? I personnaly din't use them and find the one you introduced/cleanup complete enough.

I checked with Lucas as he added them, this was a on-hold project that was not yet used, all relevant functionalities have been incorporated.

File: src/ragger/bip/path.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (lpascal-ledger) [reply src/ragger/bip/path.py]: I was more wondering why we would make a runtime check on the `format` type here and not elsewhere, as most code rely on controlled (statically checked with `mypy`) type trust.
The assert in itself is fine. You can add a more meanginful message with it:
```
assert isinstance(format, BtcDerivationPathFormat), f"Error missing value in split list {split}"
```

File: src/ragger/bip/path.py [hunk 11]
```diff
@@ -13,51 +13,36 @@
...
-    @property
-    def address_index(self) -> int:
-        return self._to_int(self[4])
+            path_bytes += int(value).to_bytes(4, byteorder='big')
+    return path_bytes
+
+
+def bitcoin_pack_derivation_path(format: BtcDerivationPathFormat, derivation_path: str) -> bytes:
+    if not isinstance(format, BtcDerivationPathFormat):
+        raise ValueError("{} must be a BtcDerivationPathFormat enum".format(format))
```
Comment (lpascal-ledger) [inline src/ragger/bip/path.py:47]: I would rather have new (Python 3.6) `f-strings` style, like already used elsewhere in `ragger` code, than older `str.format`.

Comment (fbeutin-ledger) [reply src/ragger/bip/path.py]: Done

**Review by lpascal-ledger — APPROVED** [review]

File: tests/bip/test_path.py [hunk 12]
```diff
@@ -14,7 +14,7 @@ def _test_level_n(self, n: int):
             for hardening in [False, True]:
                 hardening_char = '\'' if hardening else ''
                 previous_levels = "0/" * (n - 1)
-                path = "m/{}{}{}".format(previous_levels, variant, hardening_char)
+                path = f'm/"{previous_levels}""{variant}""{hardening_char}"'
```
Comment (lpascal-ledger) [inline tests/bip/test_path.py:17]: f'm/{previous_levels}{variant}{hardening_char}'

**Pull request merged**